python/lib/client/dmod/client/request_clients.py

```diff
@@ -1,600 +1,823 @@
+            req_params['session_secret'] = ''
+            request = MaaSDatasetManagementMessage(**req_params)
+            # TODO: (later) implement and use DmodAuthenticationFailure, though possibly down the apply_auth call stack
+            if not await self._auth_client.apply_auth(request):
+                msg = f'{self.__class__.__name__} could not apply auth to {request.__class__.__name__}'
+                raise DmodRuntimeError(msg)
+            else:
+                return request, MaaSDatasetManagementResponse
+        else:
+            return DatasetManagementMessage(**req_params), DatasetManagementResponse
+
+    async def download_dataset_item(self, dataset_name: str, item_name: str, dest: Path) -> DatasetManagementResponse:
         if dest.exists():
-            return False
+            reason = f'Destination File Exists'
+            msg = f'{self.__class__.__name__} could not download dataset item to existing path {str(dest)}'
+            return DatasetManagementResponse(success=False, reason=reason, message=msg)
+
         try:
             dest.parent.mkdir(parents=True, exist_ok=True)
         except:
-            return False
-        request = DatasetManagementMessage(action=ManagementAction.REQUEST_DATA, dataset_name=dataset_name,
-                                           data_location=item_name)
-        self.last_response: DatasetManagementResponse = await self.async_make_request(request)
+            reason = f'Unable to Create Parent Directory'
+            msg = f'{self.__class__.__name__} could not create parent directory for downloading item to {str(dest)}'
+            return DatasetManagementResponse(success=False, reason=reason, message=msg)
+
+        try:
+            request, final_response_type = self._request_prep(dataset_name=dataset_name, item_name=item_name,
+                                                              action=ManagementAction.REQUEST_DATA)
+        # TODO: (later) implement and use DmodAuthenticationFailure
+        except DmodRuntimeError as e:
+            reason = f'{self.__class__.__name__} Download Auth Failure'
+            return MaaSDatasetManagementResponse(success=False, reason=reason, message=str(e))
+
         with dest.open('w') as file:
-            for page in range(1, (self.last_response.total_pages + 1)):
-                request = DatasetManagementMessage(action=ManagementAction.DOWNLOAD_DATA, dataset_name=dataset_name,
-                                                   data_location=item_name, page=page)
-                self.last_response: DatasetManagementResponse = await self.async_make_request(request)
-                file.write(self.last_response.file_data)
+            # Do initial request outside of generator
+            await self._transport_client.async_send(data=str(request))
+            async for received_data_msg in self._transfer_receiver():
+                file.write(received_data_msg.data)
```

Comment:
It is guaranteed the full buffer passed into `write()` is flushed to the file. I think we need to loop and call write repeatedly until the full buffer has been written.

```suggestion
            async for received_data_msg in self._transfer_receiver():
                data = received_data_msg.data
                while data:
                    bytes_written = file.write(data)
                    data = data[bytes_written:]
```

Reply:
Hmm.  I'll have to spend some time reading through the `io` module docs, but I thought that use of the context manager protected against this.  I.e., that `bytes_written` would always equal `len(data)` on the first iteration of your suggested `while` loop, and that the potential risk was whether the OS actually _wrote_ things to disk (which use of the context manager ensures).  

That said, I am ok with adding this to be on the safe side.